### PR TITLE
expression: QUARTER/DATE_FORMAT compatibility with mysql for 0/0.0 values

### PR DESCRIPTION
**Pull request by ekalinin**
### What problem does this PR solve? 

Partial fix for the #11203 (+ #11223 as a global issue, + #9728)

### What is changed and how it works?

Added corner cases into `builtinCastIntAsTimeSig/builtinCastDecimalAsTimeSig`.
If the original value is 0 or 0.0 then it converted as empty Time struct (not a NULL value).

`builtinQuarterSig.evalInt` checks if we have zero date and it's not not a NULL after casting then we return 0, not NULL.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Code changes

- internal methods only

Side effects

 - Breaking backward compatibility?

Related changes

 - Need to update the documentation

Release note

 - Write release note for bug-fix.

### CODE REVIEW

**Comment by ekalinin** [comment]
TiDB now works for `0 | 0.0` as mysql:
**Schema (MySQL v5.7)**

---

**Query #1**

    select quarter(0), quarter(0.0), quarter("0000-00-00"), quarter("0000-00-00 00:00:00");

| quarter(0) | quarter(0.0) | quarter("0000-00-00") | quarter("0000-00-00 00:00:00") |
| ---------- | ------------ | --------------------- | ------------------------------ |
| 0          | 0            |                  null |                           null |

---
**Query #2**

    show warnings;

| Level   | Code | Message                                         |
| ------- | ---- | ----------------------------------------------- |
| Warning | 1292 | Incorrect datetime value: '0000-00-00'          |
| Warning | 1292 | Incorrect datetime value: '0000-00-00 00:00:00' |

---

[View on DB Fiddle](https://www.db-fiddle.com/f/26j9Zugsuw6DpMMoLKB83o/3)

File: expression/integration_test.go [hunk 1]
```diff
@@ -1280,8 +1280,15 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
...
-	result.Check(testkit.Rows("<nil> <nil> <nil> <nil> <nil> <nil>"))
+	result = tk.MustQuery(`select quarter("2012-14-20"), quarter("aa"), quarter(null), quarter(11), quarter(12.99);`)
+	result.Check(testkit.Rows("<nil> <nil> <nil> <nil> <nil>"))
+	result = tk.MustQuery(`select quarter("0000-00-00"), quarter("0000-00-00 00:00:00");`)
+	result.Check(testkit.Rows("<nil> <nil>"))
+	tk.MustQuery("show warnings").Check(testutil.RowsWithSep("|",
+		"Warning|1292|Incorrect datetime value: '0000-00-00 00:00:00.000000'",
+		"Warning|1292|Incorrect datetime value: '0000-00-00 00:00:00.000000'"))
+	result = tk.MustQuery(`select quarter(0), quarter(0.0);`)
+	result.Check(testkit.Rows("0 0"))
```
Comment (SunRunAway) [inline expression/integration_test.go]: The issue also describes the `DATE_FORMAT` problem, could you fix it together?

File: expression/builtin_time.go [hunk 2]
```diff
@@ -5535,6 +5535,15 @@ func (b *builtinQuarterSig) evalInt(row chunk.Row) (int64, bool, error) {
 	}
 
 	if date.IsZero() {
+		// MySQL compatibility, #11203
+		// 0 | 0.0 should be converted to 0 value (not null)
+		n, isNull, err := b.args[0].EvalInt(b.ctx, row)
+		isOriginalIntOrDecimalZero := n == 0 && !isNull && err == nil
```
Comment (SunRunAway) [inline expression/builtin_time.go]: 👍

File: expression/builtin_time.go [hunk 3]
```diff
@@ -5535,6 +5535,15 @@ func (b *builtinQuarterSig) evalInt(row chunk.Row) (int64, bool, error) {
...
+		// Args like "0000-00-00", "0000-00-00 00:00:00" set Fsp to 6
```
Comment (SunRunAway) [inline expression/builtin_time.go]: 👍

File: expression/integration_test.go (hunk 1 above)
Comment (ekalinin) [reply expression/integration_test.go]: Could you provide any examples for `DATE_FORMAT`?

Original issue says that `QUARTER` function returns `NULL` in tidb for 0/0.0 while mysql returns `0`.

I tried the same values for `DATE_FORMAT` and it looks like that the issue is the opposite — tidb returns `0` while mysql returns `NULL`:

**Schema (MySQL v5.7)**
```
mysql> select DATE_FORMAT(0, '%W'), DATE_FORMAT(0.0, '%W'), DATE_FORMAT('0', '%W'), DATE_FORMAT('0.0', '%W');

| DATE_FORMAT(0, '%W') | DATE_FORMAT(0.0, '%W') | DATE_FORMAT('0', '%W') | DATE_FORMAT('0.0', '%W') |
| -------------------- | ---------------------- | ---------------------- | ------------------------ |
|                 null |                   null |                   null |                     null |

mysql> show warnings;

| Level   | Code | Message                         |
| ------- | ---- | ------------------------------- |
| Warning | 1292 | Incorrect datetime value: '0'   |
| Warning | 1292 | Incorrect datetime value: '0.0' |
```

[View on DB Fiddle](https://www.db-fiddle.com/f/x4Cja21Jtv21H8qWoMdm9r/2)

**TiDB**:

```
mysql> select DATE_FORMAT(0, '%W'), DATE_FORMAT(0.0, '%W'), DATE_FORMAT('0', '%W'), DATE_FORMAT('0.0', '%W');
+----------------------+------------------------+------------------------+--------------------------+
| DATE_FORMAT(0, '%W') | DATE_FORMAT(0.0, '%W') | DATE_FORMAT('0', '%W') | DATE_FORMAT('0.0', '%W') |
+----------------------+------------------------+------------------------+--------------------------+
| 0                    | 0                      | NULL                   | NULL                     |
+----------------------+------------------------+------------------------+--------------------------+
1 row in set, 2 warnings (0.00 sec)

mysql> 
mysql> show warnings;
+---------+------+----------------------------+
| Level   | Code | Message                    |
+---------+------+----------------------------+
| Warning | 1292 | invalid time format: '0'   |
| Warning | 1292 | invalid time format: '0.0' |
+---------+------+----------------------------+
2 rows in set (0.00 sec)
```

File: expression/builtin_cast.go [hunk 4]
```diff
@@ -540,6 +540,10 @@ func (b *builtinCastIntAsTimeSig) evalTime(row chunk.Row) (res types.Time, isNul
 	if isNull || err != nil {
 		return res, isNull, err
 	}
+	// MySQL compatibility: 0.0 should not be converted to null, see #11203
+	if val == 0 {
+		return types.Time{}, false, nil
+	}
 	res, err = types.ParseTimeFromNum(b.ctx.GetSessionVars().StmtCtx, val, b.tp.Tp, int8(b.tp.Decimal))
```
[the remark this answers is not in the document]
Comment (ekalinin) [reply expression/builtin_cast.go]: Done

**Comment by ekalinin** [comment]
It looks like `DATE_FORMAT` is already fixed:

**Schema (MySQL v5.7)**
```
mysql> 
    select DATE_FORMAT(0, '%W'), 
           DATE_FORMAT(0.0, '%W'), 
           DATE_FORMAT('0', '%W'), 
           DATE_FORMAT('0.0', '%W'),
           DATE_FORMAT('0000-00-00', '%W');

| DATE_FORMAT(0, '%W') | DATE_FORMAT(0.0, '%W') | DATE_FORMAT('0', '%W') | DATE_FORMAT('0.0', '%W') | DATE_FORMAT('0000-00-00', '%W') |
| -------------------- | ---------------------- | ---------------------- | ------------------------ | ------------------------------- |
|                 NULL |                   NULL |                   NULL |                     NULL |                            NULL |

mysql>    show warnings;

| Level   | Code | Message                                |
| ------- | ---- | -------------------------------------- |
| Warning | 1292 | Incorrect datetime value: '0'          |
| Warning | 1292 | Incorrect datetime value: '0.0'        |
| Warning | 1292 | Incorrect datetime value: '0000-00-00' |
```
[View on DB Fiddle](https://www.db-fiddle.com/f/x4Cja21Jtv21H8qWoMdm9r/3)

*TiDB*:
```
mysql> select DATE_FORMAT(0, '%W'), DATE_FORMAT(0.0, '%W'), DATE_FORMAT('0', '%W'), DATE_FORMAT('0.0', '%W'), DATE_FORMAT('0000-00-00', '%W %M');
+----------------------+------------------------+------------------------+--------------------------+------------------------------------+
| DATE_FORMAT(0, '%W') | DATE_FORMAT(0.0, '%W') | DATE_FORMAT('0', '%W') | DATE_FORMAT('0.0', '%W') | DATE_FORMAT('0000-00-00', '%W %M') |
+----------------------+------------------------+------------------------+--------------------------+------------------------------------+
| NULL                 | NULL                   | NULL                   | NULL                     | NULL                               |
+----------------------+------------------------+------------------------+--------------------------+------------------------------------+
1 row in set, 5 warnings (0.00 sec)

mysql> show warnings;
+---------+------+--------------------------------------------------------+
| Level   | Code | Message                                                |
+---------+------+--------------------------------------------------------+
| Warning | 1292 | Incorrect datetime value: '0000-00-00 00:00:00'        |
| Warning | 1292 | Incorrect datetime value: '0000-00-00 00:00:00'        |
| Warning | 1292 | invalid time format: '0'                               |
| Warning | 1292 | invalid time format: '0.0'                             |
| Warning | 1292 | Incorrect datetime value: '0000-00-00 00:00:00.000000' |
+---------+------+--------------------------------------------------------+
5 rows in set (0.00 sec)
```

So i just fixed warnings for 0/0.0 values and added more tests.

**Comment by ekalinin** [comment]
PTAL

File: expression/builtin_cast.go [hunk 5]
```diff
@@ -977,6 +977,10 @@ func (b *builtinCastDecimalAsTimeSig) evalTime(row chunk.Row) (res types.Time, i
 	if isNull || err != nil {
 		return res, isNull, err
 	}
+	// MySQL compatibility: 0.0 should not be converted to null, see #11203
+	if val.IsZero() {
+		return types.Time{}, false, nil
+	}
```
Comment (SunRunAway) [inline expression/builtin_cast.go]: Could this judgement also be put into `types.ParseTimeFromFloatString`?

File: expression/integration_test.go (hunk 1 above)
Comment (SunRunAway) [reply expression/integration_test.go]: An example is `select DATE_FORMAT(0, 0);`

in TiDB, it returns NULL with a warning:
```mysql
mysql> select DATE_FORMAT(0, 0);
+-------------------+
| DATE_FORMAT(0, 0) |
+-------------------+
| NULL              |
+-------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+--------------------------------------------------------+
| Level   | Code | Message                                                |
+---------+------+--------------------------------------------------------+
| Warning | 1292 | Incorrect datetime value: '0000-00-00 00:00:00.000000' |
+---------+------+--------------------------------------------------------+
1 row in set (0.00 sec)
```

in MySQL, it returns 0 with no warning:
```mysql
mysql> select DATE_FORMAT(0, 0);
+-------------------+
| DATE_FORMAT(0, 0) |
+-------------------+
| 0                 |
+-------------------+
1 row in set (0.00 sec)
```

File: expression/builtin_cast.go (hunk 5 above)
Comment (ekalinin) [reply expression/builtin_cast.go]: Done

File: expression/integration_test.go (hunk 1 above)
Comment (ekalinin) [reply expression/integration_test.go]: Got it. Thanks.
Added a fix + test.

**Comment by ekalinin** [comment]
PTAL @XuHuaiyu

File: types/time.go [hunk 6]
```diff
@@ -1355,6 +1355,10 @@ func ParseTime(sc *stmtctx.StatementContext, str string, tp byte, fsp int8) (Tim
 
 // ParseTimeFromFloatString is similar to ParseTime, except that it's used to parse a float converted string.
 func ParseTimeFromFloatString(sc *stmtctx.StatementContext, str string, tp byte, fsp int8) (Time, error) {
+	// MySQL compatibility: 0.0 should not be converted to null, see #11203
+	if str == "0.0" {
+		return Time{}, nil
```
Comment (XuHuaiyu) [inline types/time.go]: Should this be `Time{Time: ZeroTime, Type: tp}`, too?

File: types/time.go [hunk 7]
```diff
@@ -1395,6 +1399,10 @@ func ParseDate(sc *stmtctx.StatementContext, str string) (Time, error) {
 // ParseTimeFromNum parses a formatted int64,
 // returns the value which type is tp.
 func ParseTimeFromNum(sc *stmtctx.StatementContext, num int64, tp byte, fsp int8) (Time, error) {
+	// MySQL compatibility: 0.0 should not be converted to null, see #11203
```
Comment (XuHuaiyu) [inline types/time.go]: s/ `0.0`/ 0

File: expression/builtin_time.go [hunk 8]
```diff
@@ -5540,6 +5555,15 @@ func (b *builtinQuarterSig) evalInt(row chunk.Row) (int64, bool, error) {
 	}
 
 	if date.IsZero() {
+		// MySQL compatibility, #11203
+		// 0 | 0.0 should be converted to 0 value (not null)
+		n, isNull, err := b.args[0].EvalInt(b.ctx, row)
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: `args[0]` is of type `ETDatetime` or `ETTimestamp`, we should not call `evalInt` on it.

File: expression/builtin_cast.go [hunk 9]
```diff
@@ -820,8 +820,13 @@ func (b *builtinCastRealAsTimeSig) evalTime(row chunk.Row) (types.Time, bool, er
 	if isNull || err != nil {
 		return types.Time{}, true, err
 	}
+	// MySQL compatibility: 0 should not be converted to null, see #11203
+	fv := strconv.FormatFloat(val, 'f', -1, 64)
```
Comment (XuHuaiyu) [inline expression/builtin_cast.go]: Should `builtinCastDecimalAsTimeSig` also handle this?

Comment (ekalinin) [reply expression/builtin_cast.go]: `builtinCastDecimalAsTimeSig` handles it via `ParseTimeFromFloatString`

File: types/time.go (hunk 6 above)
Comment (ekalinin) [reply types/time.go]: Fixed

File: types/time.go (hunk 7 above)
Comment (ekalinin) [reply types/time.go]: Fixed

File: expression/builtin_time.go (hunk 8 above)
Comment (ekalinin) [reply expression/builtin_time.go]: Fixed

**Comment by XuHuaiyu** [comment]
In MySQL 5.7:
``` sql
mysql> select quarter(0.00);
+---------------+
| quarter(0.00) |
+---------------+
|             0 |
+---------------+
1 row in set (0.00 sec)

mysql> select quarter(0.0);
+--------------+
| quarter(0.0) |
+--------------+
|            0 |
+--------------+
1 row in set (0.00 sec)
```

In TiDB:

``` sql
tidb> select quarter(0.00);
+---------------+
| quarter(0.00) |
+---------------+
|          NULL |
+---------------+
1 row in set, 1 warning (0.00 sec)

tidb> select quarter(0.0);
+--------------+
| quarter(0.0) |
+--------------+
|            0 |
+--------------+
1 row in set (0.00 sec)
```

**Comment by ekalinin** [comment]
`0.0*` fixed.

**Pull request merged**